### DISCUSSION TITLE
How to control the order of multiple instrumentations?

### DISCUSSION THREAD

**Discussion by seanf** [Netflix/dgs-framework#1043] [Q&A]
I want to use `MaxQueryDepthInstrumentation` and `MaxQueryComplexityInstrumentation` together, in that order. 

I have created Spring components for each of them, and also a `ChainedInstrumentation` which injects both and uses them in the desired order, but DGS seems to be picking up all three components. How can I tell DGS to use `ChainedInstrumentation` only, or alternatively can I eliminate `ChainedInstrumentation` and control the order to be sure that `MaxQueryDepthInstrumentation` will be first?

**Comment by seanf** [comment]
By reading `com.netflix.graphql.dgs.autoconfig.DgsAutoConfiguration#dgsQueryExecutor`, I have worked out that the `Instrumentation` order can be controlled using Spring's `@Order` annotation. 

In my case I put `@Order(1)` on the `MaxQueryDepthInstrumentation` component, and `@Order(2)` on the `MaxQueryComplexityInstrumentation` component (`@Priority` should work too, but I haven't tried it). 

`DgsAutoConfiguration` creates a `ChainedInstrumentation` for all the discovered `Instrumentation` components, using the `ObjectFactory` order. There doesn't appear to be any way of excluding valid `Instrumentation` components from processing by `DgsAutoConfiguration`.

**Comment by cookieMr** [comment]
Maybe this is a bit too late for @seanf but for everyone else who will read this question with a hope for an answer.

There can be only 1 spring bean of type `ChainedInstrumentation`. And that bean is already being created by `DgsAutoConfiguration` configuration class so we should not create it via any `@Configuration` class (or `@Component` annotation). And this `DgsAutoConfiguration` config class has a very interesting bean named `dgsInstrumentation`. This method takes in a (an `Optional` of) List of already ordered `Instrumentation`s. [This list is actually created by Spring by gathering all `Instrumentation`](https://docs.spring.io/spring-framework/docs/current/reference/html/core.html#beans-generics-as-qualifiers) and the order is set by `@Order` annotation of Spring (see the 1st occurrence of `@Order` in [this chapter](https://docs.spring.io/spring-framework/docs/current/reference/html/core.html#beans-autowired-annotation) - use web browser search feature).

In addition methods in `Instrumentation` are executed in a specific order.

[`@Primary` annotation](https://docs.spring.io/spring-framework/docs/current/reference/html/core.html#beans-autowired-annotation-primary) might do you a harm, as it will pick one bean over the other. Unless this is a desired behavior where the other bean is ignored (and not included into a Spring Context), do not use this `Primary` for your `Instrumentation` beans.

TL;DR

Just create additional `Instrumentation`s as pleased. Figure out proper `@Order` for all your `Instrumentation`s. Forget about `ChainedInstrumentation`. Keep in mind order of `Instrumentation`s methods execution.
